chassis.go

```diff
@@ -32,66 +32,71 @@ type Chassis struct {
 	VtepLogicalSwitches []string
 }
 
-func (odbi *ovndb) chassisAddImp(name string, hostname string, etype string, ip string) (*OvnCommand, error) {
-	// / Prepare for encap record
-	enCapUUID, err := newRowUUID()
-	if err != nil {
-		return nil, err
-	}
-	row := make(OVNRow)
-
-	if len(name) > 0 {
-		row["chassis_name"] = name
-	}
-
+func (odbi *ovndb) chassisAddImp(name string, hostname string, etype []string, ip string) (*OvnCommand, error) {
+	//Prepare for encap record
+	var encap_ids [][]string
+	var namedUUID = "named-uuid"
+	var operations []libovsdb.Operation
 	if len(etype) > 0 {
-		row["type"] = etype
-	}
+		for _, et := range etype {
+			enCapUUID, err := newRowUUID()
+			if err != nil {
+				return nil, err
+			}
+			row := make(OVNRow)
 
-	if len(ip) > 0 {
-		row["ip"] = ip
-	}
-	if uuid := odbi.getRowUUID(tableEncap, row); len(uuid) > 0 {
-		return nil, ErrorExist
+			if len(name) > 0 {
+				row["chassis_name"] = name
+			}
+			var encap_id []string
+			encap_id = append(encap_id, namedUUID)
+			encap_id = append(encap_id, enCapUUID)
+			encap_ids = append(encap_ids, encap_id)
+			if len(ip) > 0 {
+				row["ip"] = ip
+			}
+			row["type"] = et
+			if uuid := odbi.getRowUUID(tableEncap, row); len(uuid) > 0 {
+				return nil, ErrorExist
+			}
+			insertEncapOp := libovsdb.Operation{
+				Op:       opInsert,
+				Table:    tableEncap,
+				Row:      row,
+				UUIDName: enCapUUID,
+			}
+			// check if chassis already exists with that encap
+			if uuid := odbi.getRowUUID(tableChassis, row); len(uuid) > 0 {
+				return nil, ErrorExist
+			}
+			operations = append(operations, insertEncapOp)
+		}
 	}
 
-	insertEncapOp := libovsdb.Operation{
-		Op:       opInsert,
-		Table:    tableEncap,
-		Row:      row,
-		UUIDName: enCapUUID,
-	}
 	// Prepare for chassis record
 	ChassisUUID, err := newRowUUID()
 	if err != nil {
 		return nil, err
 	}
-
 	rowChassis := make(OVNRow)
-
-	var encap_id []string
-	var namedUUID = "named-uuid"
-	encap_id = append(encap_id, namedUUID)
-	encap_id = append(encap_id, enCapUUID)
-
-	rowChassis["encaps"] = encap_id
+	// consolidate encaps
+	var encaps []interface{}
+	encaps = append(encaps, "set" )
+	encaps = append(encaps, encap_ids)
+	rowChassis["encaps"] = encaps
 	if len(name) > 0 {
```

Comment:
"name" is mandatory column for Chassis. It should error out if len(name) == 0.

Reply:
I added validations for all mandatory args